evmlib/src/external_signer.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2024 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::common::{Address, Amount, Calldata, QuoteHash, QuotePayment, U256};
+use crate::contract::data_payments::{DataPaymentsHandler, MAX_TRANSFERS_PER_TRANSACTION};
+use crate::contract::network_token::NetworkToken;
+use crate::contract::{data_payments, network_token};
+use crate::utils::http_provider;
+use crate::Network;
+use std::collections::HashMap;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Network token contract error: {0}")]
+    NetworkTokenContract(#[from] network_token::Error),
+    #[error("Chunk payments contract error: {0}")]
+    ChunkPaymentsContract(#[from] data_payments::error::Error),
+}
+
+#[allow(dead_code)]
+/// Approve an address / smart contract to spend this wallet's payment tokens.
+///
+/// Returns the transaction calldata (input, to).
+pub fn approve_to_spend_tokens_calldata(
+    network: &Network,
+    spender: Address,
+    value: U256,
+) -> (Calldata, Address) {
+    let provider = http_provider(network.rpc_url().clone());
+    let network_token = NetworkToken::new(*network.payment_token_address(), provider);
+    network_token.approve_calldata(spender, value)
+}
+
+#[allow(dead_code)]
+/// Transfer payment tokens from the supplied wallet to an address.
+///
+/// Returns the transaction calldata (input, to).
+pub fn transfer_tokens_calldata(
+    network: &Network,
+    receiver: Address,
+    amount: U256,
+) -> (Calldata, Address) {
+    let provider = http_provider(network.rpc_url().clone());
+    let network_token = NetworkToken::new(*network.payment_token_address(), provider);
+    network_token.transfer_calldata(receiver, amount)
+}
+
+#[allow(dead_code)]
+pub struct PayForQuotesCalldataReturnType {
+    pub batched_calldata_map: HashMap<Calldata, Vec<QuoteHash>>,
+    pub to: Address,
+    pub approve_spender: Address,
+    pub approve_amount: Amount,
+}
+
+#[allow(dead_code)]
+/// Use this wallet to pay for chunks in batched transfer transactions.
+/// If the amount of transfers is more than one transaction can contain, the transfers will be split up over multiple transactions.
+///
+/// Returns PayForQuotesCalldataReturnType, containing calldata of the transaction batches along with the approval details for the spender.
```

Comment:
What exactly is 'call data'?

Reply:
I have used calldata as a general term to refer to the data that is needed to form a transaction, including transaction input and the to address. `Calldata` the type is an alias to `Bytes` and should be set as the input of a transaction.